app/scripts/migrations/081.js

```diff
@@ -0,0 +1,68 @@
+import { cloneDeep } from 'lodash';
+import {
+  WALLET_SNAP_PERMISSION_KEY,
+  SnapCaveatType,
+} from '@metamask/rpc-methods';
+
+const version = 81;
+
+/**
+ * Prior to this migration, snap <> dapp permissions were wildcards i.e. `wallet_snap_*`.
+ * Now the permission has been changed to `wallet_snap` and the current snap permissions
+ * that are under wildcards will be added as caveats to a parent `wallet_snap` permission.
+ */
+export default {
+  version,
+  async migrate(originalVersionedData) {
+    const versionedData = cloneDeep(originalVersionedData);
+    versionedData.meta.version = version;
+    const state = versionedData.data;
+    const newState = transformState(state);
+    versionedData.data = newState;
+    return versionedData;
+  },
+};
+
+function transformState(state) {
+  const PermissionController = state?.PermissionController || {};
+
+  const { subjects } = PermissionController;
+
+  const snapPrefix = 'wallet_snap_';
```

Comment:
Can we detect snaps earlier an bail out of this entire thing if we don't see any snaps permissions? Maybe, we should just bail if the build type is not Flask?

Reply:
Yeah, true.

Reply:
On second thought, introducing `isFlask` adds some inconsistency with tests, i.e. the tests will fail on non-flask builds. Alternatively, the cost to grab all the permission keys at once and check for snap permissions is not really much different than current implementation. I'm going to keep implementation as is.